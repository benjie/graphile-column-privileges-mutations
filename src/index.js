import express from 'express';
import { postgraphile } from 'postgraphile';
import { NodePlugin } from 'graphile-build';
// import { MutationPlugin } from 'graphile-build';
import PgMutationUpdateDeletePlugin from './plugins/PgMutationUpdateDeletePlugin';
import PgMutationCreatePlugin from './plugins/PgMutationCreatePlugin';
import PgSimplifyInflectorPlugin from './plugins/PgSimplifyInflectorPlugin';
const app = express();

app.use(
  postgraphile('postgres://localhost:5432/mutation_example', 'public', {
    graphiql: true,
    enhanceGraphiql: true,
    enableCors: true,
    // skipPlugins: [NodePlugin],
    dynamicJson: true,
    appendPlugins: [
      PgSimplifyInflectorPlugin,
      PgMutationCreatePlugin,
      PgMutationUpdateDeletePlugin
    ],
    graphileBuildOptions: {
      // disable the default mutations
      pgDisableDefaultMutations: true
    }
  })
);

export default app;
