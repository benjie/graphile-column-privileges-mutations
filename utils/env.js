import { cleanEnv, str, port } from 'envalid';

export default cleanEnv(
  process.env,
  {
    PGUSER: str({ default: 'postgres' }),
    PGHOST: str({ default: 'localhost' }),
    PGDATABASE: str({ default: 'mutation_example' }),
    SCHEMA: str({ default: 'app_public' }),
    PGPASSWORD: str({ default: 'password' }),
    PGPORT: port({ default: 5432 })
  },
  { dotEnvPath: null }
);
