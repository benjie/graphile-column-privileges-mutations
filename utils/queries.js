import gql from 'graphql-tag';

export const CreateUserMutation = gql`
  mutation CreateUserMutation(
    $email: String!
    $username: String!
    $password: String!
  ) {
    createUser(
      input: {
        user: { email: $email, username: $username, password: $password }
      }
    ) {
      user {
        id
        username
        email
      }
    }
  }
`;

export const CreateUserMutationBad = gql`
  mutation CreateUserMutation(
    $email: String!
    $username: String!
    $password: String!
  ) {
    createUser(
      input: {
        user: { email: $email, username: $username, password: $password }
      }
    ) {
      user {
        id
        username
        email
        password
      }
    }
  }
`;

export const DeleteUserMutation = gql`
  mutation DeleteUserMutation($username: String!) {
    deleteUserByUsername(input: { username: $username }) {
      clientMutationId
    }
  }
`;

export const DeleteUserMutationBad = gql`
  mutation DeleteUserMutation($username: String!) {
    deleteUserByUsername(input: { username: $username }) {
      user {
        id
      }
    }
  }
`;

export const UpdateUserMutation = gql`
  mutation UpdateUserMutation(
    $username: String!
    $newname: String
    $password: String
  ) {
    updateUserByUsername(
      input: {
        userPatch: { username: $newname, password: $password }
        username: $username
      }
    ) {
      user {
        id
        email
        username
      }
    }
  }
`;

export const UpdateUserMutationBad = gql`
  mutation UpdateUserMutation(
    $username: String!
    $newname: String
    $password: String
  ) {
    updateUserByUsername(
      input: {
        userPatch: { username: $newname, password: $password }
        username: $username
      }
    ) {
      user {
        id
        email
        username
        password
      }
    }
  }
`;
