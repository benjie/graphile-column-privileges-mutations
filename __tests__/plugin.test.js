import { GraphQLTest } from '../utils/graphql';
import { CreateUserMutation, CreateUserMutationBad } from '../utils/queries';
import env from '../utils/env';
import { snapshot } from '../utils/clean';
import { PgMutationCreatePlugin, PgMutationUpdateDeletePlugin } from '../src';
import { Schema } from 'pg-query-string';

const { SCHEMA } = env;
const schema = new Schema(SCHEMA);
const table = schema.table('users', {
  id: 'int',
  username: 'text',
  email: 'text',
  password: 'text'
});

const getDbString = () =>
  `postgres://${env.PGUSER}:${env.PGPASSWORD}@${env.PGHOST}:${env.PGPORT}/${env.PGDATABASE}`;

const { setup, teardown, graphQL } = GraphQLTest(
  {
    appendPlugins: [PgMutationUpdateDeletePlugin, PgMutationCreatePlugin],
    graphileBuildOptions: {
      // disable the default mutations
      pgDisableDefaultMutations: true
    },
    schema: SCHEMA,
    graphqlRoute: '/graphql'
  },
  getDbString()
);

beforeAll(async () => {
  await setup();
});
afterAll(async () => {
  await teardown();
});

it('allows select without password', async () => {
  await graphQL(async (query, pgClient) => {
    const { data } = await query(CreateUserMutation, {
      email: 'pyramation@example.com',
      username: 'pyramation',
      password: 'password'
    });

    expect(snapshot(data)).toMatchSnapshot();

    // ensure password exists
    await pgClient.query("select set_config('role', $1, true)", ['postgres']);
    const {
      rows: [row]
    } = await pgClient.query(
      table.select(['id', 'email', 'username', 'password'], {
        username: 'pyramation'
      })
    );
    expect(snapshot(row)).toMatchSnapshot();
  });
});

it('disallows password on select', async () => {
  await graphQL(async query => {
    const { data, errors } = await query(CreateUserMutationBad, {
      email: 'pyramation@example.com',
      username: 'pyramation',
      password: 'password'
    });
    expect(errors).toBeTruthy();
    expect(errors[0].message).toMatchSnapshot();
  });
});
