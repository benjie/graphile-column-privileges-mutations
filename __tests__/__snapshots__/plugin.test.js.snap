// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create error 1`] = `"permission denied for table users"`;

exports[`create success 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "user": Object {
        "email": "pyramation@example.com",
        "id": "[ID]",
        "username": "pyramation",
      },
    },
  },
}
`;

exports[`create success 2`] = `
Object {
  "email": "pyramation@example.com",
  "id": "[ID]",
  "password": "password",
  "username": "pyramation",
}
`;

exports[`delete error 1`] = `"Cannot query field \\"user\\" on type \\"DeleteUserPayload\\"."`;

exports[`delete success 1`] = `
Object {
  "data": Object {
    "deleteUserByUsername": Object {
      "clientMutationId": null,
    },
  },
}
`;

exports[`delete success 2`] = `undefined`;

exports[`update error 1`] = `"permission denied for table users"`;

exports[`update success 1`] = `
Object {
  "data": Object {
    "updateUserByUsername": Object {
      "user": Object {
        "email": "updateme@example.com",
        "id": "[ID]",
        "username": "newname",
      },
    },
  },
}
`;

exports[`update success 2`] = `
Object {
  "email": "updateme@example.com",
  "id": "[ID]",
  "password": "newpassword",
  "username": "newname",
}
`;
